### DISCUSSION TITLE
Making the preset configuration even simpler

### DISCUSSION THREAD

**Discussion by faassen** [alvis/presetter#7] [Ideas]
Let's consider this structure inside a preset:

```javascript
{
    template: {
      '.babelrc.json': resolve(TEMPLATES, 'babelrc.yaml'),
      '.eslintrc.json': resolve(TEMPLATES, 'eslintrc.yaml'),
      '.gitignore': resolve(TEMPLATES, 'gitignore'),
      '.jestrc.json': resolve(TEMPLATES, 'jestrc.yaml'),
      '.npmignore': resolve(TEMPLATES, 'npmignore'),
      '.prettierrc.json': resolve(TEMPLATES, 'prettierrc.yaml'),
      'tsconfig.json': resolve(TEMPLATES, 'tsconfig.yaml'),
      'tsconfig.build.json': resolve(TEMPLATES, 'tsconfig.build.yaml'),
    },
    noSymlinks: ['.gitignore'],
    scripts: resolve(TEMPLATES, 'scripts.yaml'),
}
```

The invocation of resolve is repeated. I am wondering why the framework can't do it for us, if at least we made the templates directory available with a separate `templates` entry. That way it could look like this:

```javascript
{
    templatesPath: resolve(__dirname, '..', 'templates'),
    template: {
      '.babelrc.json': 'babelrc.yaml',
      '.eslintrc.json': 'eslintrc.yaml',
      '.gitignore': 'gitignore',
      '.jestrc.json': 'jestrc.yaml',
      '.npmignore': 'npmignore',
      '.prettierrc.json': 'prettierrc.yaml',
      'tsconfig.json': 'tsconfig.yaml',
      'tsconfig.build.json': 'tsconfig.build.yaml',
    },
    noSymlinks: ['.gitignore'],
    scripts: 'scripts.yaml',
}
```

That's less typing, but another phenomenon presents itself: this mapping could be automated using a convention over configuration pattern:

```javascript
{
    templatesPath: resolve(__dirname, '..', 'templates'),
    templates: ['.babelrc.json', '.eslintrc.json', '.gitignore', '.jestrc.json', '.npmignore', '.prettierrc.json', 'tsconfig.json', 'tsconfig.build.json']
    noSymlinks: ['.gitignore'],
}
```

The rule would be: remove the initial underscore, and look for the templates under `templatesPath`. Use the extension (`yaml`, '.json', etc) to determine what merge rule to use. For scripts we always look for `scripts.*` under templates so we can leave this out.

To still allow complete control we could allow exceptions:

```javascript
// earlier
const TEMPLATES = resolve(__dirname, '..', 'templates'); 
// 
{
    templatesPath: TEMPLATES
    templates: ['.babelrc.json', {path: '.eslintrc.json', template: resolve(TEMPLATES,  'my-own-eslintrc.yaml')}, '.gitignore', '.jestrc.json', '.npmignore', '.prettierrc.json', 'tsconfig.json', 'tsconfig.build.json']
    noSymlinks: ['.gitignore'],
}
```

I suspect that 'noSymlinks' could also be part of this structure, and the `template` entry could become optional, so we could do:

```javascript
{
    templatesPath: TEMPLATES
    templates: ['.babelrc.json', '.eslintrc.json', {path: "gitignore", noSymlinks: true}, '.jestrc.json', '.npmignore', '.prettierrc.json', 'tsconfig.json', 'tsconfig.build.json']
}
```

Using convention over configuration here would encourage a consistent naming pattern for templates, which is I think desirable in a preset. And it's simply less typing and less to think about.

There are questions to resolve concerning `supplementaryConfigs` and the like.

**Comment by alvis** [comment]
In principle, I quite like the last form -- a list file paths to export with the flexibility to supply more for some exceptional cases.

However, I'm thinking about a potential conflict with `templatesPath` (and likewise `supplementaryConfigsPath`) when `templates` is not just a map of template files, but the actual content of the templates.

For example, in the case of persetter-preset-rollup:
https://github.com/alvis/presetter/blob/70ac3865fccb823f8ad962609dd9d88e83b504a4/packages/preset-rollup/source/index.ts#L53-L63

`rollup.config.ts` is a ts file dynamically generated and it cannot be generated like an ordinary template with some static `variables`. Here, the template variable `rollupImport` is a variable generated based on other static `variables`.

Then, if the `template` object must also be able to take the literal content, `templatesPath` will become redundant.
May be a workaround is having the container path as the first element of the list, following another list of templates?
e.g.
```ts
{
  templates: [
    resolve(__dirname, '..', 'templates'),
    ['.babelrc.json', '.eslintrc.json', {path: "gitignore", noSymlinks: true}, '.jestrc.json', '.npmignore', '.prettierrc.json', 'tsconfig.json', 'tsconfig.build.json']
  ]
}
```

and continue to accept the classic form for advanced use cases?
```ts
{
  templates: {
    '.babelrc.json': {...}
  }
}
```

**Reply by faassen** [reply]
I'm not sure I fully understand your rollup example or why you should want to make a nested array structure like you propose. But my general response is to offer more hooks to do things.

If we want to be able to completely control how the template is converted to the file:

```javascript
{
   templates: ['.babelrc.json', { path: 'rollup.config.ts', render: (template: string) => doWhatever(template) }
}
```

So presetter still loads the file using its usual rules, but then leaves processing its content up to you. This would be implemented as the actual lower layer in the software, and the standard convention would be a function that we pass as `render` during preprocessing.

If we just want to have a dynamically interpolated variable but otherwise use the standard processing system (if that works for strings), we could do that too:

```javascript
{
  templates: ['.babelrc.json', { path: 'rollup.config.ts', additionalVariables: (variables) => doSomething(variables) }
}
```

Here `additionalVariables` would get the variables passed in and then return an object with additional derived variables, which will then be passed into the normal variable interpolation logic.

**Reply by alvis** [reply]
Absorbing your suggestion, I'm considering to make presetter to accept following format as well:
```ts
{
  templates: [
    resolve(__dirname, '..', 'templates'),
    [
      '.babelrc.json',
      {name: 'gitignore', noSymlinks: true},
      {name: 'rollup.config.ts', resolveVariables:(context) => ({...'a map of variables to be used in rollup.config.ts'})},
     ...
    ]
  ]
}
```

The first field of `templates` indicates the root of all templates.
Then the second field contains a list of template file names, with some special cases as suggested.

**Reply by faassen** [reply]
I'm not a big fan of using array tuples when we have an object key available. So I prefer having a separate templateRoot key in the outer configuration, rather than dealing with a 2 element array.

**Reply by alvis** [reply]
Hmm. My gut feeling is mixed. 🤔 On one hand, just needed to input the template file names, not full paths is better. But on the other hand, havng a `templateRoot` key will break the current format, which I would strongly be against because it's the way needed internally to link the templates and the final output, also a map will ensure that there'll be no duplicated entries.
